### PR TITLE
Fix some threading things and add additional thread unittests

### PR DESCRIPTION
**Pull request by kwolekr**
- Fix thread name reset on start()
- Fully reset thread state on kill()
- Add unittests to check for correct object states under various circumstances

Just making a PR for the CI checks :-)

### CODE REVIEW

File: src/threading/thread.cpp [hunk 1]
```diff
@@ -156,21 +155,30 @@ bool Thread::stop()
...
 
 
-void Thread::wait()
+bool Thread::wait()
 {
-	if (!m_running)
-		return;
+	MutexAutoLock lock(m_mutex);
+
+	if (!m_joinable)
```
Comment (est31) [inline src/threading/thread.cpp]: As we have the mutex now, we can keep the name for the variable to be m_running, right?

**Comment by est31** [comment]
@SmallJoker can you test whether this PR makes it compile for you?

File: src/threading/thread.h [hunk 2]
```diff
@@ -81,9 +81,9 @@ class Thread {
 	/*
 	 * Waits for thread to finish.
 	 * Note:  This does not stop a thread, you have to do this on your own.
-	 * Returns immediately if the thread is not started.
+	 * Returns immediately if the thread is not started or has already finished.
```
Comment (est31) [inline src/threading/thread.h]: Perhaps also point out what it returns?

File: src/threading/thread.cpp (hunk 1 above)
Comment (kwolekr) [inline src/threading/thread.cpp]: No, it still needs to be there.
A thread can't be joined more than once.

Comment (est31) [inline src/threading/thread.cpp]: That's why the lock is there. once the thread is joined, nobody else can obtain the lock. So a check for m_running is sufficient, no?

File: src/threading/thread.h (hunk 2 above)
Comment (kwolekr) [inline src/threading/thread.h]: Okay.

**Pull request closed without merging**